File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    N = len(topping)
    a_arr = [0] * 10001
    b_arr = [0] * 10001
    a_set = set()
    b_set = set()
    
    a_arr[topping[0]] = 1
    a_set.add(topping[0])
    for i in range(1, N):
        b_arr[topping[i]] += 1
        b_set.add(topping[i])
    
    if len(a_set) == len(b_set):
        answer += 1
        
    for i in range(1, N-1):
        a_arr[topping[i]] += 1
        a_set.add(topping[i])
        b_arr[topping[i]] -= 1
        if b_arr[topping[i]] == 0:
            b_set.remove(topping[i])
            
        if len(a_set) == len(b_set):
            answer += 1
    
    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.Map;
import java.util.HashMap;
class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> bro_1 = new HashMap<>();
        Map<Integer, Integer> bro_2 = new HashMap<>();
        
        for (int e : topping) {
            bro_2.put(e, bro_2.getOrDefault(e, 0) + 1);
        }
        
        for (int e : topping) {
            bro_1.put(e, bro_1.getOrDefault(e, 0) + 1);
            
            if (bro_2.get(e) - 1 == 0) {
                bro_2.remove(e);
            } else {
                bro_2.put(e, bro_2.get(e) - 1);
            }
            
            if (bro_1.size() == bro_2.size()) {
                answer++;
            }
        }
        return answer;
    }
}